File: greeter_client.cc
#include <iostream>
#include <thread>
#include <memory>
#include <string>
#include <experimental/random>

#include <grpcpp/grpcpp.h>

#include "helloworld.grpc.pb.h"

using grpc::Channel;
using grpc::ChannelArguments;
using grpc::ClientContext;
using grpc::ClientReader;
using grpc::Status;
using helloworld::Empty;
using helloworld::CcuReply;
using helloworld::Server;


class Client 
{
    public:
        Client(std::shared_ptr<Channel> channel) : stub_(Server::NewStub(channel)) {};

    ClientContext context;

    void StreamCcu() 
    {
        Empty request;
        CcuReply reply;

        std::unique_ptr<ClientReader<CcuReply> > reader(stub_->StreamCcu(&context, request));
        while (reader->Read(&reply)) {
            std::cout << "Greeter received: " << reply.n() << std::endl;
        }
        Status status = reader->Finish();
    }

    private:
        std::unique_ptr<Server::Stub> stub_;
};


int main(int argc, char** argv)
{
    ChannelArguments args;
    Client client(grpc::CreateCustomChannel("localhost6:50051", grpc::InsecureChannelCredentials(), args));
  
    std::this_thread::sleep_for(std::chrono::seconds(std::experimental::randint(1,5)));
    
    std::thread t([](Client *srv) {
        srv->StreamCcu(); 
    }, &client);
    
    std::this_thread::sleep_for(std::chrono::seconds(std::experimental::randint(1,5)));
    
    client.context.TryCancel();
    t.join();
    return 0;
}

File: greeter_client.py
import random
import time
import threading
import logging

import grpc

import helloworld_pb2
import helloworld_pb2_grpc

class Client:
    
    def __init__(self):
        channel = grpc.insecure_channel('localhost:50051')
        self.stub = helloworld_pb2_grpc.ServerStub(channel)

    def run(self):
        threading.Thread(target=self.__stream_ccu, daemon=True).start()

    def __stream_ccu(self):
        responses = self.stub.StreamCcu(helloworld_pb2.Empty())
        for response in responses:
            print("Greeter client received: %d" % response.n)
    
if __name__ == '__main__':
    logging.basicConfig()
    client = Client()
    time.sleep(random.randint(1,5))
    client.run()
    time.sleep(random.randint(1,5))
